autogen/agentchat/groupchat.py

```diff
@@ -30,45 +41,64 @@ def agent_by_name(self, name: str) -> Agent:
         """Find the next speaker based on the message."""
         return self.agents[self.agent_names.index(name)]
 
-    def next_agent(self, agent: Agent) -> Agent:
+    def next_agent(self, agent: Agent, agents: List[Agent]) -> Agent:
         """Return the next agent in the list."""
-        return self.agents[(self.agent_names.index(agent.name) + 1) % len(self.agents)]
-
-    def select_speaker_msg(self):
+        if agents == self.agents:
+            return agents[(self.agent_names.index(agent.name) + 1) % len(agents)]
+        else:
+            offset = self.agent_names.index(agent.name) + 1
+            for i in range(len(self.agents)):
+                if self.agents[(offset + i) % len(self.agents)] in agents:
+                    return self.agents[(offset + i) % len(self.agents)]
+
+    def select_speaker_msg(self, agents: List[Agent]):
         """Return the message for selecting the next speaker."""
         return f"""You are in a role play game. The following roles are available:
 {self._participant_roles()}.
 
 Read the following conversation.
-Then select the next role from {self.agent_names} to play. Only return the role."""
+Then select the next role from {[agent.name for agent in agents]} to play. Only return the role."""
 
     def select_speaker(self, last_speaker: Agent, selector: ConversableAgent):
         """Select the next speaker."""
-        selector.update_system_message(self.select_speaker_msg())
-
-        # Warn if GroupChat is underpopulated, without established changing behavior
-        n_agents = len(self.agent_names)
-        if n_agents < 3:
-            logger.warning(
-                f"GroupChat is underpopulated with {n_agents} agents. Direct communication would be more efficient."
-            )
-
+        if self.func_call_filter and self.messages and "function_call" in self.messages[-1]:
+            # find agents with the right function_map which contains the function name
+            agents = [
+                agent for agent in self.agents if agent.can_execute_function(self.messages[-1]["function_call"]["name"])
+            ]
+            if len(agents) == 1:
+                # only one agent can execute the function
+                return agents[0]
+            elif not agents:
+                raise ValueError(
+                    f"No agent can execute the function {self.messages[-1]['name']}. "
+                    "Please check the function_map of the agents."
+                )
```

Comment:
If no function name matches, another possible reason is that the LLM has some mistakes in calling the function. Shall we raise the error directly, or return an error message to the chat?
(`function_exec` will return a "Error: Function {func_name} not found" if no function name matches.)

Reply:
Thanks for suggesting this scenario. Addressed this scenario.